File: src_python/lattice/tr7to5.py
import numpy as np
import qnnum as qnn
import qnndarray as qna
import qnvec as qnv
import cython

def tr7to5e(ei:qna.QnNdarray) -> qna.QnNdarray:  # for direct space vector
  #      for dodecagonal QCs
  sd =[[0, 1, 0, 0, 0,-1, 0],
       [1, 0, 1, 0, 0, 0, 0],
       [0, 1, 0, 1, 0, 0, 0],
       [0, 0, 1, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 1]]
  shape=ei.shape
  nv=shape[0]
  fi = qna.zeros((nv,5))
  qn0=qnn.zero()
  for i in range(nv):
    for  j in range(5):
      fi[i][j]=qn0
      for k in range(7):
        fi[i][j]+=ei[i][k]*sd[j][k]
  return fi

def tr7to5i(ei:qna.QnNdarray) -> qna.QnNdarray:  # for reciprocal vector
  #      for dodecagonal QCs
  sd = [[1, 0, 0, 0,-1, 0, 0],
        [0, 1, 0, 0, 0,-1, 0],
        [0, 0, 1, 0, 1, 0, 0],
        [0, 0, 0, 1, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1]]
  
  shape=ei.shape
  nv=shape[0]
  qn0=qnn.zero()
  fi = qna.zeros((nv,5))
  for i in range(nv):
    for  j in range(5):
      fi[i][j]=qn0
      for  k in range(7):
        fi[i][j]+=ei[i][k]*sd[j][k]
  return fi

File: src_python/dihed/lattice/fsrc/tr7to5.f90
subroutine tr7to5e(ei,fi)
    !       for dodecagonal
    real fi(*),ei(*)
    integer sd(7,7)
    data sd / 0, 1, 0, 0, 0,-1, 0,&
      &          1, 0, 1, 0, 0, 0, 0,&
      &          0, 1, 0, 1, 0, 0, 0,&
      &          0, 0, 1, 0, 1, 0, 0,&
      &          1, 0,-1, 0, 1, 0, 0,&
      &          0, 1, 0,-1, 0, 1, 0,&
      &          0, 0, 0, 0, 0, 0, 1/

    do j=1,6
      fi(j)=0.
      if(j<=4) then
        do k=1,6
          fi(j)=fi(j)+sd(k,j)*ei(k)
           !     fi(j)=fi(j)+sd(k,j)*ei(k)/3
        end do
      else if (j==5) then
        fi(5)=ei(7)
        fi(6)=0
      end if
    end do
    return
  END subroutine tr7to5e

  subroutine tr7to5i(ei,fi)
    !       for dodecagonal
    integer fi(*),ei(*)
    integer sd(7,7)
    data sd / 1, 0, 0, 0,-1, 0, 0,&
      &          0, 1, 0, 0, 0,-1, 0,&
      &          0, 0, 1, 0, 1, 0, 0,&
      &          0, 0, 0, 1, 0, 1, 0,&
      &          1, 0,-1, 0, 1, 0, 0,&
      &          0, 1, 0,-1, 0, 1, 0,&
      &          0, 0, 0, 0, 0, 0, 1/
  
    do j=1,6
      fi(j)=0.
      if(j<=4) then
        do k=1,6
          fi(j)=fi(j)+sd(k,j)*ei(k)
        end do
      else if (j==5) then
        fi(5)=ei(7)
        fi(6)=0
      end if
    end do
    return
  END subroutine tr7to5i
